aws/data_source_aws_instances.go

```diff
@@ -0,0 +1,106 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceAwsInstances() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsInstancesRead,
+
+		Schema: map[string]*schema.Schema{
+			"filter":        dataSourceFiltersSchema(),
+			"instance_tags": tagsSchemaComputed(),
+
+			"ids": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"private_ips": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"public_ips": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func dataSourceAwsInstancesRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	filters, filtersOk := d.GetOk("filter")
+	tags, tagsOk := d.GetOk("instance_tags")
+
+	if !filtersOk && !tagsOk {
+		return fmt.Errorf("One of filters or instance_tags must be assigned")
+	}
+
+	params := &ec2.DescribeInstancesInput{}
+	if filtersOk {
+		params.Filters = buildAwsDataSourceFilters(filters.(*schema.Set))
+	}
+	if tagsOk {
+		params.Filters = append(params.Filters, buildEC2TagFilterList(
+			tagsFromMap(tags.(map[string]interface{})),
+		)...)
+	}
+
+	log.Printf("[INFO] Describing EC2 instances: %s", params)
+
+	var instanceIds, privateIps, publicIps []string
+	err := conn.DescribeInstancesPages(params, func(resp *ec2.DescribeInstancesOutput, isLast bool) bool {
+		// loop through reservations, and remove terminated instances, populate instance slice
```

Comment:
Option: could always use a `filter` on `instance-state-name` and append user filters to it, to save you this looping and possibly entire API pagination calls

- http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html

Reply:
Ha, very valid point, I'll change that 👍